### ISSUE TITLE
Compression Issue?

### ISSUE THREAD

**Issue by Maruhl** [MiloszKrajewski/K4os.Compression.LZ4#87]
**Description**
Compression at "L12_MAX" weird - or in general?
Or I use it completely wrong.

I have attached the file I am trying to compress "Globals.zip".
[Globals.zip](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/files/12358506/Globals.zip)

**To reproduce**
```
private static void CompressionTest()
{
	using var input = File.OpenRead(@"Globals.lsf");
	using var output = File.Create(@"Globals.lsf.lz4");

	using LZ4EncoderStream encodedStream = LZ4Stream.Encode(output, LZ4Level.L12_MAX);
	input.CopyTo(encodedStream);
}
```

**Expected behavior**
Significantly smaller file or binary data.

**Actual behavior**
Doesn't really seem to compress everything.

**Environment**
 - CPU: AMD
 - OS: Windows 10 x64
 - .NET: Net 7.0
 - LZ4: K4os.Compression.LZ4.Streams (1.3.6)

**Additional context**
I have additionally looked at the files, not even the strings are compressed. Something is compressed, but not everything.

HexView of the plain file:
![grafik](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/assets/41585619/fe309fa7-6cdd-44ac-922a-d1c44768c3b2)

HexView of the compressed file:
![grafik](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/assets/41585619/214c938c-26f0-4788-befb-c9b223cb93cb)

**Comment by MiloszKrajewski** [comment]
Does original `lz4.exe` (https://github.com/lz4/lz4/releases/tag/v1.9.4) compress any better? 
If so, then we might have a bug here, if not then this is just related to LZ4 as algorithm.

Some notes though:
LZ4 is *fast* compression algorithm, not high compression algorithm nor general purpose compression algorithm. It uses dictionary compression, so its compression depends on repeated data.
Your comment about "not even the strings are compressed" suggest that you expected Huffman coding (which is quite common and relatively slow) but it is not part of LZ4 algorithm.

So, does `lz4.exe` produce different results?

**Comment by Maruhl** [comment]
Hi thanks for the quick reply.

I am trying to replace a part of this library (https://github.com/Norbyte/lslib) that works with a native lz4 implementation.

The native implementation is here:
https://github.com/Norbyte/lslib/tree/master/LSLibNative/lz4

I have no idea about compression algorithms, at least a method is called that ends with "HC", probably "Huffman Coding" is meant.

Now when I use lz4 directly, the result is almost the same, so it won't be a bug.

```
lz4.exe -z -12 Globals.lsf

Compressed filename will be : Globals.lsf.lz4
Compressed 20856981 bytes into 18077530 bytes ==> 86.67%
```

But what I find interesting is that this is decompressed correctly. Only not compressed again.

The compressed file:
[Globals.lsf.compressed.zip](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/files/12366362/Globals.lsf.compressed.zip)

Basically, I just want to decompress the data and compress it again as if it were original - at least close to it.

**Comment by MiloszKrajewski** [comment]
> I have no idea about compression algorithms, at least a method is called that ends with "HC", probably "Huffman Coding" is meant.

Never thought about this wat of reading "HC", but no, HC here stands for High Compression.

> Now when I use lz4 directly, the result is almost the same, so it won't be a bug.
> 
> ```
> lz4.exe -z -12 Globals.lsf
> 
> Compressed filename will be : Globals.lsf.lz4
> Compressed 20856981 bytes into 18077530 bytes ==> 86.67%
> ```

Ok, so it is as expected then.

> But what I find interesting is that this is decompressed correctly. Only not compressed again.

This is also expected. There are more settings than just compression level, like block size, checksums, etc. So files might be slightly different, but compatible.

> Basically, I just want to decompress the data and compress it again as if it were original - at least close to it.

So yes, this is the way to do it.
